pkg/controller/istiocsr/controller.go

```diff
@@ -0,0 +1,341 @@
+package istiocsr
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+
+	v1alpha1 "github.com/openshift/cert-manager-operator/api/operator/v1alpha1"
+)
+
+var (
+	// requestEnqueueLabelKey is the label key name used for filtering reconcile
+	// events to include only the resources created by the controller.
+	requestEnqueueLabelKey = "app"
+
+	// requestEnqueueLabelKey is the label value used for filtering reconcile
+	// events to include only the resources created by the controller.
+	requestEnqueueLabelValue = "cert-manager-istio-csr"
+)
+
+// Reconciler reconciles a IstioCSR object
+type Reconciler struct {
+	Client
+
+	ctx           context.Context
+	eventRecorder record.EventRecorder
+	log           logr.Logger
+	scheme        *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=istiocsrs,verbs=get;list;watch
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=istiocsrs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=istiocsrs/finalizers,verbs=update
```

Comment:
IIRC from our earlier discussions, deletion of the resources are manual by the user:
if we add `finalizers` here and the controller never cleans up the resource,
how can users proceed with a safe deletion (without getting stuck)?